scheduler/scheduler.go

```diff
@@ -112,21 +121,89 @@ func Schedule(ctx context.Context, cmd data.ScheduledCommand) error {
 	return err
 }
 
-// ScheduleFromString schedules a command using its string representation.
-func ScheduleFromString(ctx context.Context, commandString string, etc data.ScheduledCommand) error {
+// Schedule registers a data.ScheduledCommand with the scheduler so it will be
+// requested appropriately.
+func Schedule(ctx context.Context, cmd data.ScheduledCommand) (int64, error) {
+	err := auth.CheckPermissions(ctx, cmd.UserName, cmd.Command, cmd.CommandEntry)
+	if err != nil {
+		return 0, err
+	}
+	da, err := dataaccess.Get()
+	if err != nil {
+		return 0, err
+	}
+	err = da.ScheduleCreate(ctx, &cmd)
+	if err != nil {
+		return 0, err
+	}
+
+	return cmd.ScheduleID, schedule(cmd)
+}
 
+// ScheduleFromString schedules a command using its string representation.
+func ScheduleFromString(ctx context.Context, commandString string, etc data.ScheduledCommand) (int64, error) {
 	tokens, err := command.Tokenize(commandString)
 	if err != nil {
-		return err
+		return 0, err
 	}
 
 	cmdEntry, cmdInput, err := retrieval.CommandFromTokensByName(ctx, tokens)
 	if err != nil {
-		return err
+		return 0, err
 	}
 
+	cmdInput.Original = commandString
+
 	etc.CommandEntry = *cmdEntry
 	etc.Command = cmdInput
 
 	return Schedule(ctx, etc)
 }
+
+// Cancel cancels the scheduled command with the given id. Even in the event of
+// an error, the command may still be cancelled.
+func Cancel(ctx context.Context, scheduleID int64) error {
+	da, err := dataaccess.Get()
+	if err != nil {
+		return err
+	}
+
+	err = da.ScheduleDelete(ctx, scheduleID)
```

Comment:
Idiomatically I'd expect to see something more like:

```go
        err = da.ScheduleDelete(ctx, scheduleID)
	if err != nil {
		return err
	}

	err = cron.RemoveByTag(fmt.Sprintf("%d", scheduleID))
	if err2 != nil {
		return err
	}
```

Reply:
I was thinking that we might still want to end the job even if it fails to be removed from `da`, but maybe creating that kind of inconsistency is worse.